Mathlib/Data/Bitvec/Lemmas.lean

```diff
@@ -124,4 +124,44 @@ theorem ofFin_toFin {n} (v : BitVec n) : ofFin (toFin v) = v := by
   rfl
 #align bitvec.of_fin_to_fin Std.BitVec.ofFin_toFin
 
+/-!
+  ## Distributivity of ofFin
+  We add simp-lemmas that show how `ofFin` distributes over various bitvector operations, showing
+  that bitvector operations are equivalent to `Fin` operations
+-/
+@[simp] lemma neg_ofFin (x : Fin (2^w)) : -(ofFin x) = ofFin (-x) := by
+  rw [neg_eq_zero_sub]; rfl
+
+@[simp] lemma ofFin_and_ofFin (x y : Fin (2^w)) : ofFin x &&& ofFin y = ofFin (x &&& y) := rfl
+@[simp] lemma ofFin_or_ofFin  (x y : Fin (2^w)) : ofFin x ||| ofFin y = ofFin (x ||| y) := rfl
+@[simp] lemma ofFin_xor_ofFin (x y : Fin (2^w)) : ofFin x ^^^ ofFin y = ofFin (x ^^^ y) := rfl
+@[simp] lemma ofFin_add_ofFin (x y : Fin (2^w)) : ofFin x + ofFin y   = ofFin (x + y)   := rfl
+@[simp] lemma ofFin_sub_ofFin (x y : Fin (2^w)) : ofFin x - ofFin y   = ofFin (x - y)   := rfl
+@[simp] lemma ofFin_mul_ofFin (x y : Fin (2^w)) : ofFin x * ofFin y   = ofFin (x * y)   := rfl
+
+/-!
+## Ring
+-/
+
+lemma zero_eq_ofFin_zero : 0#w = ofFin 0 := rfl
+lemma one_eq_ofFin_one   : 1#w = ofFin 1 := rfl
+
+/-! Now we can define an instance of `CommRing (BitVector w)` straightforwardly in terms of the
+    existing instance `CommRing (Fin (2^w))` -/
+instance : CommRing (BitVec w) where
+  add_assoc       := by intro ⟨_⟩ ⟨_⟩ ⟨_⟩; simp [add_assoc]
+  zero_add        := by intro ⟨_⟩; simp [zero_eq_ofFin_zero]
+  add_zero        := by intro ⟨_⟩; simp [zero_eq_ofFin_zero]
+  sub_eq_add_neg  := by intro ⟨_⟩ ⟨_⟩; simp [sub_eq_add_neg]
+  add_comm        := by intro ⟨_⟩ ⟨_⟩; simp [add_comm]
+  left_distrib    := by intro ⟨_⟩ ⟨_⟩ ⟨_⟩; simp [left_distrib]
+  right_distrib   := by intro ⟨_⟩ ⟨_⟩ ⟨_⟩; simp [right_distrib]
+  zero_mul        := by intro ⟨_⟩; simp [zero_eq_ofFin_zero]
+  mul_zero        := by intro ⟨_⟩; simp [zero_eq_ofFin_zero]
+  mul_assoc       := by intro ⟨_⟩ ⟨_⟩ ⟨_⟩; simp [mul_assoc]
+  mul_comm        := by intro ⟨_⟩ ⟨_⟩; simp [mul_comm]
+  one_mul         := by intro ⟨_⟩; simp [one_eq_ofFin_one]
+  mul_one         := by intro ⟨_⟩; simp [one_eq_ofFin_one]
+  add_left_neg    := by intro ⟨_⟩; simp [zero_eq_ofFin_zero]
```

Comment:
I think it would be better to use `Function.Injective.commRing`.

Reply:
That seems cleaner, yes. However, `Function.Injective.commRing` requires instances of `[SMul ℕ β] [SMul ℤ β] [Pow β ℕ]`, which `CommRing` by itself does not need.
Is this by design, it seems `Function.Injective.commRing` shouldn't need them if it's just defining an instance of `CommRing`?

Reply:
> which CommRing by itself does not need.

Actually it does need them, but it generates them automatically if you forget them.
The point here is that probably you don't want the auto-generated ones; the manual implementations in terms of `Fin` will likely have better properties.